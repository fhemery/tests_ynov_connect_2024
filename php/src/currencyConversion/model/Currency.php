<?php

namespace kata\model;

use kata\external\CurrencyIsoCode;

enum Currency
{
    case Euro;
    case Dollar;
    case Pound;
    case Yen;

    public function toString() {
        return match($this) {
            self::Euro => "€",
            self::Dollar => "$",
            self::Pound => "£",
            self::Yen => "¥"
        };
    }

    public function toIsoCurrency(): CurrencyIsoCode
    {
        return match($this) {
            self::Euro => CurrencyIsoCode::EUR,
            self::Dollar => CurrencyIsoCode::USD,
            self::Pound => CurrencyIsoCode::GBP,
            self::Yen => CurrencyIsoCode::JPY,
        };
    }
}