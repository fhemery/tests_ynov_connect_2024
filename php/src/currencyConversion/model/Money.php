<?php

namespace kata\model;

class Money
{

    public float $amount;
    public Currency $currency;

    public function __construct(float $amount, Currency $currency)
    {
        $this->amount = $amount;
        $this->currency = $currency;
    }
}