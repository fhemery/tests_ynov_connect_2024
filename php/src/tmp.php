$apiMock = $this->createMock(ConversionRateApi::class);
$apiMock->expects($this->once())
->method('getRate')
->with(Currency::Dollar->toIsoCurrency(), Currency::Euro->toIsoCurrency())
->willReturn(1.5);

$converter = new CurrencyConverter($apiMock);

//  This call does not work. A test double is probably necessary
$result = $converter->sum(Currency::Euro, [new Money(1, Currency::Dollar)]);

$this->assertEquals(1.5, $result->amount);
$this->assertEquals(Currency::Euro, $result->currency);