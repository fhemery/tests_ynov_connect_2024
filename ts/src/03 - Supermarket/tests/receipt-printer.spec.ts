import {ReceiptPrinter} from "../src/ReceiptPrinter";
import {Receipt} from "../src/model/Receipt";
import {Product} from "../src/model/Product";
import {ProductUnit} from "../src/model/ProductUnit";

function extractFromTicket(ticket: string, lineNumber: number, fromCol: number, toCol: number): string {
    return ticket.split('\n')[lineNumber].substring(fromCol, toCol);
}

describe('ReceiptPrinter', () => {
    let receiptPrinter: ReceiptPrinter;

    beforeEach(() => {
        receiptPrinter = new ReceiptPrinter();
    });

    it('should print a receipt with 0 product', () => {
        const receipt = new Receipt();

        const ticket = receiptPrinter.printReceipt(receipt);

        expect(ticket).toEqual('\nTotal:                              0.00');
        expect(ticket).toContain('Total:');
        expect(ticket).toContain('0.00');
    })

    describe('with one product', () => {
        it('should print the correct receipt (with helper function)', () => {
            const receipt = new Receipt();
            const apples = new Product('apples', ProductUnit.Kilo);
            receipt.addProduct(apples, 2, 2.5, 5);

            const ticket = receiptPrinter.printReceipt(receipt);

            expect(extractFromTicket(ticket, 0, 0, 6)).toBe('apples');
            expect(extractFromTicket(ticket, 0, 36, 40)).toBe('5.00');
        });

        it('should print the correct receipt (with snapshot)', () => {
            const receipt = new Receipt();
            const apples = new Product('apples', ProductUnit.Kilo);
            receipt.addProduct(apples, 2, 2.5, 5);

            const ticket = receiptPrinter.printReceipt(receipt);

            expect(ticket).toMatchSnapshot();
        });

    });
});
