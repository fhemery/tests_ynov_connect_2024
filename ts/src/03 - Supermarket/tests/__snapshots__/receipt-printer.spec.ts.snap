// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiptPrinter with one product should print the correct receipt (with snapshot) 1`] = `
"apples                              5.00
  2.50 * 2.000

Total:                              5.00"
`;
